palace/linalg/hypre.cpp

```diff
@@ -47,6 +47,31 @@ HypreCSRMatrix::HypreCSRMatrix(hypre_CSRMatrix *mat) : mat(mat)
   width = hypre_CSRMatrixNumCols(mat);
 }
 
+HypreCSRMatrix::HypreCSRMatrix(mfem::SparseMatrix &m)
+  : palace::Operator(m.Height(), m.Width())
+{
+  // Initialize the internal hypre matrix.
```

Comment:
Hypre's memory location always matches MFEM's (Hypre uses device when MFEM does). So, we can do this without modifying the Hypre matrix's memory location.

```suggestion
  mat = hypre_CSRMatrixCreate(height, width, m.NumNonZeroElems());
  hypre_CSRMatrixI(mat) = m.ReadWriteI();
  hypre_CSRMatrixJ(mat) = m.ReadWriteJ();
  hypre_CSRMatrixData(mat) = m.ReadWriteData();
  hypre_CSRMatrixInitialize(mat);
```

Also, `hypre_CSRMatrixDestroy` always calls 'free` on the I pointer, so I think you either need to set a flag and set it to NULL in the`HypreCSRMatrix` destructor before calling the Hypre free destroy function, or just explicitly copy into it. You need to also set `hypre_CSRMatrixOwnsData(mat) = 0` here.

Reply:
I would rather only have the owning construction then, as that avoids the need for any flag, or creating a new array and explicit copying of the contents of `I`. To construct from a `SparseMatrix` would always require `std::move` and the incoming `SparseMatrix` would then continue to point to the data arrays, but no longer own the data. Works with `*LoseMat()` or with `SpMat()` then, but `LoseMat` seems more expressive. See https://github.com/awslabs/palace/pull/217/commits/3d05f08ff404af4451d49a78d1c022d0200dec4e

Reply:
The flag is already there in `hypre_CSRMatrixOwnsData(mat)` isn't it? So it would just involve checking the value of the flag in the destructor and setting ` hypre_CSRMatrixI(mat) = nullptr` if false.

I like and suggested the non-owning approach since it's consistent with the `HypreVector` above. I'm not sure I understand the point about always needing `std::move`. Your non-owning constructor can take a (const) reference to a `mfem::SparseMatrix` and just wrap the data pointers. Then you can call this constructor with `BilinearForm::SpMat()` without any transfer of ownership.

Reply:
The transfer of ownership creates issues since MFEM and Hypre may be using different memory allocators. In general we want to have the same library which is allocating memory be the same one which is destroying it to avoid problems down the road. For example, consider MFEM using a managed memory backend but Hypre not, or vice versa. The libraries might be doing something more sophisticated with memory allocation/deallocation and assume that when you are using the libraries' memory deallocation routines, you have used the corresponding allocation routines.